### PR TITLE
TW-1827: online status is not updated correctly

### PR DESCRIPTION
**Pull request by sherlockvn**
### Updated:
- for chat we don't have `lastActiveTimestamp`, keep `offine` status (in this case, it means user never online before or in the case of can't get presence of user)
- for people that offline longer than 10 hours, we display `aWhile ago` instead of `offline`

### android:

https://github.com/linagora/twake-on-matrix/assets/43041967/12bc31cc-babd-4fb2-a860-bcd8297ddc32

### ios:

https://github.com/linagora/twake-on-matrix/assets/43041967/6e10b9f4-b82f-4bda-b9f4-935748262cc6

### CODE REVIEW

**Comment by Te-Z** [comment]
@tprudentova is this solution ok for you ?

> ### Updated:
> * for chat we don't have `lastActiveTimestamp`, keep `offine` status (in this case, it means user never online before or in the case of can't get presence of user)
> * for people that offline longer than 10 hours, we display `aWhile ago` instead of `offline`

**Comment by tprudentova** [comment]
@Te-Z if the user was online >10 hours ago, but we can get information about the time when that person was online, please display exact time (like "online 3 days ago")
"Offline" and "online a while ago" only for the users for which we don't have any information

**Comment by sherlockvn** [comment]
> @Te-Z if the user was online >10 hours ago, but we can get information about the time when that person was online, please display exact time (like "online 3 days ago") "Offline" and "online a while ago" only for the users for which we don't have any information

After how many days offline should the status change to "online a while ago"?. For users we don't have information, we display "offline"

**Comment by Te-Z** [comment]
> > @Te-Z if the user was online >10 hours ago, but we can get information about the time when that person was online, please display exact time (like "online 3 days ago") "Offline" and "online a while ago" only for the users for which we don't have any information
> 
> After how many days offline should the status change to "online a while ago"?. For users we don't have information, we display "offline"

@sherlockvn please tag @tprudentova if you need to ask her or send her a message via chat for more efficiency

**Comment by sherlockvn** [comment]
@tprudentova

**Comment by tprudentova** [comment]
@sherlockvn @Te-Z I think we can display "online a while ago" after 30 days of absence, similar to telegram

**Comment by sherlockvn** [comment]
okay np

File: lib/utils/date_time_extension.dart [hunk 1]
```diff
@@ -160,9 +160,14 @@ extension DateTimeExtension on DateTime {
...
-  bool isLessThanTenHoursAgo({DateTime? other}) {
+  bool isLessThanADayAgo({DateTime? other}) {
     other ??= DateTime.now();
-    return other.difference(this) < const Duration(hours: 10);
+    return other.difference(this) < const Duration(hours: 24);
+  }
+
+  bool isLessThan30DaysAgo({DateTime? other}) {
+    other ??= DateTime.now();
+    return other.difference(this) < const Duration(days: 30);
```
Comment (hieutbui) [inline lib/utils/date_time_extension.dart:170]: Don't forget to write test for these functions!

Comment (sherlockvn) [reply lib/utils/date_time_extension.dart:170]: done

File: test/utils/get_localize_status_test.dart [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+
+      final textWidgetFinder = find.byKey(textWidgetKey);
+
+      expect(textWidgetFinder, findsOneWidget);
+
+      final Text textWidget = tester.widget(textWidgetFinder) as Text;
+
+      expect(textWidget.data, isNotNull);
+
+      expect(textWidget.data, equals('online 20h ago'));
```
Comment (hieutbui) [inline test/utils/get_localize_status_test.dart:260]: All tests have similar steps, I think you can create a common function. WDYT?

Comment (sherlockvn) [reply test/utils/get_localize_status_test.dart]: okay i create a function for WHEN part, other part still the same to keep the format of the test for easier to follow

**Review by hieutbui — APPROVED** [review]

File: lib/pages/chat/chat.dart [hunk 3]
```diff
@@ -258,6 +260,42 @@ class ChatController extends State<Chat>
   SuggestionsController<Map<String, String?>> suggestionsController =
       SuggestionsController();
 
+  ValueNotifier<CachedPresence?> cachedPresenceNotifier = ValueNotifier(null);
+
+  final StreamController<ConnectivityResult>
+      connectivityResultStreamController = StreamController.broadcast();
```
Comment (Te-Z) [inline lib/pages/chat/chat.dart:266]: this controller is not used nor disposed

File: lib/pages/chat/chat.dart [hunk 4]
```diff
@@ -258,6 +260,42 @@ class ChatController extends State<Chat>
   SuggestionsController<Map<String, String?>> suggestionsController =
       SuggestionsController();
 
+  ValueNotifier<CachedPresence?> cachedPresenceNotifier = ValueNotifier(null);
```
Comment (Te-Z) [inline lib/pages/chat/chat.dart:263]: you forgot to dispose it

File: lib/pages/chat/chat.dart (hunk 3 above)
Comment (nqhhdev) [reply lib/pages/chat/chat.dart:266]: Pls remove it

Comment (sherlockvn) [reply lib/pages/chat/chat.dart]: done

**Review by hieutbui — APPROVED** [review]

**Review by Te-Z — APPROVED** [review]

**Pull request merged**